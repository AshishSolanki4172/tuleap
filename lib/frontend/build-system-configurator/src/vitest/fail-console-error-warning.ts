/**
 * Copyright (c) Enalean, 2022-Present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

import { beforeEach, afterEach, vi } from "vitest";

let is_console_error_or_warning = false;

beforeEach(() => {
    is_console_error_or_warning = false;
    const originalError = global.console.error;
    vi.spyOn(global.console, "error").mockImplementation((...args: unknown[]) => {
        is_console_error_or_warning = true;
        originalError(...args);
    });
    const originalWarn = global.console.warn;
    vi.spyOn(global.console, "warn").mockImplementation((...args: unknown[]) => {
        is_console_error_or_warning = true;
        originalWarn(...args);
    });
});

afterEach(() => {
    if (is_console_error_or_warning) {
        throw new Error("Console warnings and errors are not allowed");
    }
});
