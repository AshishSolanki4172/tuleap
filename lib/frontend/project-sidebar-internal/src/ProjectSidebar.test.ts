/**
 * Copyright (c) 2021-Present Enalean
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import { describe, it, expect } from "vitest";
import ProjectSidebar from "./ProjectSidebar.vue";
import { shallowMount } from "@vue/test-utils";
import { example_config } from "./project-sidebar-example-config";
import SidebarCollapseButton from "./SidebarCollapseButton.vue";

describe("ProjectSidebar", () => {
    it("displays sidebar", () => {
        const wrapper = shallowMount(ProjectSidebar, {
            props: { config: JSON.stringify(example_config) },
        });

        expect(wrapper.element).toMatchSnapshot();
    });

    it("displays nothing when config attribute is incorrect", () => {
        const wrapper = shallowMount(ProjectSidebar, { props: { config: "" } });

        expect(wrapper.element.childNodes).toHaveLength(0);
    });

    it.each([
        [false, false],
        [true, true],
    ])(
        `Given configuration does not contain information about is_collapsible,
        And collapsed attribute is %s
        Then sidebar can be collapsed and is currently collapsed = %s`,
        (collapsed, expected) => {
            // eslint-disable-next-line @typescript-eslint/no-unused-vars
            const { is_collapsible, ...config_without_is_collapsible } = example_config;

            const wrapper = shallowMount(ProjectSidebar, {
                props: { config: JSON.stringify(config_without_is_collapsible), collapsed },
            });

            expect(wrapper.element.classList.contains("sidebar-collapsed")).toBe(expected);
            expect(wrapper.findComponent(SidebarCollapseButton).props("is_sidebar_collapsed")).toBe(
                expected,
            );
            expect(
                wrapper.findComponent(SidebarCollapseButton).props("can_sidebar_be_collapsed"),
            ).toBe(true);
        },
    );

    it.each([
        [false, false],
        [true, true],
    ])(
        `Given configuration have is_collapsible = true,
        And collapsed attribute is %s
        Then sidebar can be collapsed and is currently collapsed = %s`,
        (collapsed, expected) => {
            const wrapper = shallowMount(ProjectSidebar, {
                props: {
                    config: JSON.stringify({ ...example_config, is_collapsible: true }),
                    collapsed,
                },
            });

            expect(wrapper.element.classList.contains("sidebar-collapsed")).toBe(expected);
            expect(wrapper.findComponent(SidebarCollapseButton).props("is_sidebar_collapsed")).toBe(
                expected,
            );
            expect(
                wrapper.findComponent(SidebarCollapseButton).props("can_sidebar_be_collapsed"),
            ).toBe(true);
        },
    );

    it.each([[false], [true]])(
        `Given configuration have is_collapsible = false,
        And collapsed attribute is %s
        Then sidebar cannot be collapsed and is currently not collapsed`,
        (collapsed) => {
            const wrapper = shallowMount(ProjectSidebar, {
                props: {
                    config: JSON.stringify({ ...example_config, is_collapsible: false }),
                    collapsed,
                },
            });

            expect(wrapper.element.classList.contains("sidebar-collapsed")).toBe(false);
            expect(wrapper.findComponent(SidebarCollapseButton).props("is_sidebar_collapsed")).toBe(
                false,
            );
            expect(
                wrapper.findComponent(SidebarCollapseButton).props("can_sidebar_be_collapsed"),
            ).toBe(false);
        },
    );
});
