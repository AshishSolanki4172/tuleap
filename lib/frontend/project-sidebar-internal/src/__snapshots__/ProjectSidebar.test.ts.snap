// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProjectSidebar > displays sidebar 1`] = `
<aside
  class="sidebar"
  data-v-73c4a8c1=""
>
  <sidebar-logo-stub
    data-v-73c4a8c1=""
  />
  <div
    class="sidebar-content-vertical-scroll"
    data-v-73c4a8c1=""
  >
    <sidebar-header-stub
      data-v-73c4a8c1=""
      is_sidebar_collapsed="false"
    />
    <tool-list-stub
      data-v-73c4a8c1=""
    />
    <sidebar-collapse-button-stub
      can_sidebar_be_collapsed="true"
      data-v-73c4a8c1=""
      is_sidebar_collapsed="false"
    />
  </div>
  <div
    class="sidebar-spacer"
    data-v-73c4a8c1=""
  />
  <sidebar-footer-stub
    data-v-73c4a8c1=""
  />
</aside>
`;
